File: built-in plugins dsp/mb-dist-soft.cpp
/**
 * MB Soft Clipper
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Gentle soft clipping
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DIST_SOFT_H
#define MB_DIST_SOFT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDistSoft : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-dist-soft";
    static constexpr const char* PLUGIN_NAME    = "MB Soft Clipper";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -6f;  // range [-24, 0]
    float knee = 0.5f;  // range [0, 1]
    };

    MbDistSoft() = default;
    ~MbDistSoft() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -24f, 0f);
        params.knee = std::clamp(params.knee, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Soft Clipper
        return input;
    }
};

#endif // MB_DIST_SOFT_H

File: built-in plugins dsp/mb-dist-soft.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbDistSoftPlugin: PluginDefinition = { id: 'mb-dist-soft', slug: 'mb-dist-soft', name: 'MB Soft Clipper', category: 'effect', type: 'distortion', version: '1.0.0', description: 'Gentle soft clipping', author: 'Max Booster', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -6, minValue: -24, maxValue: 0, automatable: true }, { id: 'knee', name: 'Knee', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { threshold: -6, knee: 0.5 } };

export default MbDistSoftPlugin;
